metrics/cmd/heapster-apiserver/app/server.go

```diff
@@ -0,0 +1,149 @@
+package app
+
+import (
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	"k8s.io/heapster/metrics/options"
+	"k8s.io/kubernetes/pkg/admission"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/apiserver/authenticator"
+	"k8s.io/kubernetes/pkg/controller/framework/informers"
+	"k8s.io/kubernetes/pkg/genericapiserver"
+	genericauthorizer "k8s.io/kubernetes/pkg/genericapiserver/authorizer"
+	genericoptions "k8s.io/kubernetes/pkg/genericapiserver/options"
+	"k8s.io/kubernetes/pkg/registry/cachesize"
+)
+
+// NewAPIServerCommand creates a *cobra.Command object with default parameters
+func NewAPIServerCommand() *cobra.Command {
+	s := genericoptions.NewServerRunOptions()
+	s.AddUniversalFlags(pflag.CommandLine)
+	cmd := &cobra.Command{
+		Use:  "heapster-apiserver",
+		Long: `heapster apiserver`,
+		Run: func(cmd *cobra.Command, args []string) {
+		},
+	}
+
+	return cmd
+}
+
+type HeapsterAPIServer struct {
+	*genericapiserver.GenericAPIServer
+	options *options.HeapsterRunOptions
+}
+
+// Run runs the specified APIServer. This should never exit.
+func (h *HeapsterAPIServer) RunServer() error {
+	h.Run(h.options.ServerRunOptions)
+	return nil
+}
+
+func NewHeapsterApiServer(s *options.HeapsterRunOptions) (*HeapsterAPIServer, error) {
+	server, err := newAPIServer(s.ServerRunOptions)
+	if err != nil {
+		return &HeapsterAPIServer{}, err
+	}
+	return &HeapsterAPIServer{
+		GenericAPIServer: server,
+		options:          s,
+	}, nil
+}
+
+func newAPIServer(s *genericoptions.ServerRunOptions) (*genericapiserver.GenericAPIServer, error) {
+	genericapiserver.DefaultAndValidateRunOptions(s)
+
+	resourceConfig := genericapiserver.NewResourceConfig()
+
+	storageGroupsToEncodingVersion, err := s.StorageGroupsToEncodingVersion()
+	if err != nil {
+		glog.Fatalf("error generating storage version map: %s", err)
+	}
+	storageFactory, err := genericapiserver.BuildDefaultStorageFactory(
+		s.StorageConfig, s.DefaultStorageMediaType, api.Codecs,
+		genericapiserver.NewDefaultResourceEncodingConfig(), storageGroupsToEncodingVersion,
+		[]unversioned.GroupVersionResource{}, resourceConfig, s.RuntimeConfig)
+	if err != nil {
+		glog.Fatalf("error in initializing storage factory: %s", err)
+	}
+
+	authz, err := authenticator.New(authenticator.AuthenticatorConfig{
+		BasicAuthFile:     s.BasicAuthFile,
+		ClientCAFile:      s.ClientCAFile,
+		TokenAuthFile:     s.TokenAuthFile,
+		OIDCIssuerURL:     s.OIDCIssuerURL,
+		OIDCClientID:      s.OIDCClientID,
+		OIDCCAFile:        s.OIDCCAFile,
+		OIDCUsernameClaim: s.OIDCUsernameClaim,
+		OIDCGroupsClaim:   s.OIDCGroupsClaim,
+		KeystoneURL:       s.KeystoneURL,
+	})
+	if err != nil {
+		glog.Fatalf("Invalid Authentication Config: %v", err)
+	}
+
+	authorizationModeNames := strings.Split(s.AuthorizationMode, ",")
+	authorizationConfig := genericauthorizer.AuthorizationConfig{
```

Comment:
Should this have separate authorization, or just bounce off of the main API server use SARs, etc?

cc @deads2k